### PR TITLE
dnf: allow handling of modules without a default profile

### LINKED ISSUE #56504: dnf fails to install module streams unless both stream and profile are named
##### SUMMARY
I am trying to  install a module on RHEL8 and/or Fedora29 and fail to do so unless I pass both the stream and profile.
Depending on the syntax used the module either errors out or says that the module is already installed.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
dnf
##### ANSIBLE VERSION

```paste below
# ansible --version
ansible 2.9.0.dev0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /tmp/ansible-git/lib/ansible
  executable location = /tmp/ansible-git/bin/ansible
  python version = 2.7.15 (default, Oct 15 2018, 15:26:09) [GCC 8.2.1 20180801 (Red Hat 8.2.1-2)]
```

also happens with
```
% ansible --version
ansible 2.7.10
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/egolov/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.7.3 (default, Mar 27 2019, 13:41:07) [GCC 8.3.1 20190223 (Red Hat 8.3.1-2)]
```
##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT

Fedora 29 with both Python2 and Python3 on the control node
Fedora 29 and RHEL 8 on the target

##### STEPS TO REPRODUCE

The documentation of the `dnf` module suggests you can use the following syntax to install a module:
* `name: '@postgresql:9.6/client'`
* `name: '@postgresql:9.6'`
* `name: '@postgresql/client'`

Looking at the code, just `@postgresql` should also work.

```
# ansible -m dnf -a 'name=@hub' localhost
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: TypeError: 'in <string>' requires string as left operand, not NoneType
localhost | FAILED! => {
    "changed": false, 
    "module_stderr": "Traceback (most recent call last):\n  File \"/root/.ansible/tmp/ansible-tmp-1557992375.73-85400188950236/AnsiballZ_dnf.py\", line 125, in <module>\n    _ansiballz_main()\n  File \"/root/.ansible/tmp/ansible-tmp-1557992375.73-85400188950236/AnsiballZ_dnf.py\", line 117, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/root/.ansible/tmp/ansible-tmp-1557992375.73-85400188950236/AnsiballZ_dnf.py\", line 54, in invoke_module\n    imp.load_module('__main__', mod, module, MOD_DESC)\n  File \"/tmp/ansible_dnf_payload_ziVgHS/__main__.py\", line 1293, in <module>\n  File \"/tmp/ansible_dnf_payload_ziVgHS/__main__.py\", line 1282, in main\n  File \"/tmp/ansible_dnf_payload_ziVgHS/__main__.py\", line 1261, in run\n  File \"/tmp/ansible_dnf_payload_ziVgHS/__main__.py\", line 932, in ensure\n  File \"/tmp/ansible_dnf_payload_ziVgHS/__main__.py\", line 865, in _is_module_installed\nTypeError: 'in <string>' requires string as left operand, not NoneType\n", 
    "module_stdout": "", 
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", 
    "rc": 1
}

# ansible -m dnf -a 'name=@hub/default' localhost
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: TypeError: 'in <string>' requires string as left operand, not NoneType
localhost | FAILED! => {
    "changed": false, 
    "module_stderr": "Traceback (most recent call last):\n  File \"/root/.ansible/tmp/ansible-tmp-1557992389.11-263418060033161/AnsiballZ_dnf.py\", line 125, in <module>\n    _ansiballz_main()\n  File \"/root/.ansible/tmp/ansible-tmp-1557992389.11-263418060033161/AnsiballZ_dnf.py\", line 117, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/root/.ansible/tmp/ansible-tmp-1557992389.11-263418060033161/AnsiballZ_dnf.py\", line 54, in invoke_module\n    imp.load_module('__main__', mod, module, MOD_DESC)\n  File \"/tmp/ansible_dnf_payload_a6AtAR/__main__.py\", line 1293, in <module>\n  File \"/tmp/ansible_dnf_payload_a6AtAR/__main__.py\", line 1282, in main\n  File \"/tmp/ansible_dnf_payload_a6AtAR/__main__.py\", line 1261, in run\n  File \"/tmp/ansible_dnf_payload_a6AtAR/__main__.py\", line 932, in ensure\n  File \"/tmp/ansible_dnf_payload_a6AtAR/__main__.py\", line 865, in _is_module_installed\nTypeError: 'in <string>' requires string as left operand, not NoneType\n", 
    "module_stdout": "", 
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", 
    "rc": 1
}

# ansible -m dnf -a 'name=@hub:pre-release' localhost
localhost | SUCCESS => {
    "changed": false, 
    "msg": "Nothing to do", 
    "rc": 0, 
    "results": [
        "Module hub:pre-release installed."
    ]
}

# ansible -m dnf -a 'name=@hub:pre-release/default' localhost
localhost | CHANGED => {
    "changed": true, 
    "msg": "", 
    "rc": 0, 
    "results": [
        "Module hub:pre-release/default installed.", 
        "Installed: hub-prerelease-8.module_1978+3584bea7.x86_64"
    ]
}

```

##### EXPECTED RESULTS
As the `@module` syntax is not documented, I would not *expect* it to work, but it would be really nice if it did.
For `@module:stream` and `@module/profile` I would expect them to install the module, just as `@module:stream/profile` did.

##### ACTUAL RESULTS
See above

### PR DESCRIPTION
**Pull request by evgeni**
##### SUMMARY

DNF modules can be shipped without a default profile. In this case
calling 'install module:stream' errors out, as DNF doesn't know
what to install (as this is determined by the profile).

One solution is to explicitly call 'install module:stream/profile'.
However, it can be beneficial to only enable a module, without
installing anything from it. This is useful when the module is just
needed for dependencies, and maybe doesn't ship *any* profiles at all.

Examples for such modules are pki-deps and pki-core in EL8.

Fixes: #56504

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

dnf

##### ADDITIONAL INFORMATION

On EL8, the `pki-core` module can't be *installed*, just *enabled*, as it does not have any profiles.

```paste below
# ansible -m dnf -a 'name=@pki-core:10.6' localhost
localhost | FAILED! => {
    "changed": false,
    "failures": [
        "pki-core:10.6 Problems in request:\nbroken groups or modules: pki-core:10.6"
    ],
    "msg": "Failed to install some of the specified packages",
    "rc": 1,
    "results": []
}

# dnf module install pki-core:10.6
Failed to set locale, defaulting to C.UTF-8
Last metadata expiration check: 19:38:41 ago on Wed Nov 10 12:54:13 2021.
No profiles for module pki-core:10.6
Error: Problems in request:
broken groups or modules: pki-core:10.6

# ansible -m dnf -a 'name=@pki-core:10.6 state=enabled' localhost   
localhost | CHANGED => {
    "changed": true,
    "msg": "",
    "rc": 0,
    "results": [
        "Module pki-core:10.6 enabled."
    ]
}
```

### CODE REVIEW

**Review by s-hertel — CHANGES_REQUESTED** [review]
I see why this felt like a natural extension to the already-existing module support, but I think the initial feature also would have been better as a separate module for maintainability. It seems like there's consensus from https://github.com/ansible/ansible/issues/64852 (including from the active maintainer at the time who added the initial support) is that it should be a separate module. 

Another thing to note is that it would break feature parity with the `yum` module that was added in https://github.com/ansible/ansible/commit/6d95624c22d1c275974db1e7c94dcc171dc1457e.

New modules should generally be opened against a community collection. There are some guidelines here: https://docs.ansible.com/ansible/latest/community/contributing_maintained_collections.html#deciding-where-your-contribution-belongs

**Comment by evgeni** [comment]
Thanks for the quick review!

> I see why this felt like a natural extension to the already-existing module support, but I think the initial feature also would have been better as a separate module for maintainability. It seems like there's consensus from #64852 (including from the active maintainer at the time who added the initial support) is that it should be a separate module.

I do agree that *all of* modularity support should never have been added to the `dnf` module to begin with, but in an own `dnf_module` module (how often will I write module in this post?). As right now, `dnf` can handle *some* modules, but not others, and this makes users unhappy.

> Another thing to note is that it would break feature parity with the `yum` module that was added in [6d95624](https://github.com/ansible/ansible/commit/6d95624c22d1c275974db1e7c94dcc171dc1457e).

Given the `dnf` module can handle installing/uninstalling modules, and `yum` cannot, makes this argument moot, no? :)

Also, there are gems like
https://github.com/ansible/ansible/blob/382a35323400e841caa0560132779f0ab563ad2e/lib/ansible/modules/dnf.py#L1405-L1408
As `yum` and `dnf` (the tools, not the modules) don't have feature parity anyways.

> New modules should generally be opened against a community collection. There are some guidelines here: https://docs.ansible.com/ansible/latest/community/contributing_maintained_collections.html#deciding-where-your-contribution-belongs

Placing `dnf_module` into a collection, would also mean that users of `ansible-core` (and other Ansible distribututions that don't ship the `community.*` collections) would get the `dnf` module that can do a bit of modularity, but not `dnf_module` which can do the other "half", which sounds wrong to me. I think what I am trying to say is: either all of this needs to be in the `dnf` module, or we need to rip out the current modularity support out of it, and place all into `dnf_module`, thus making it the only one handling modules. Otherwise we end up with terrible UX :(

[this post, without the quotes, includes 16 times the word "module"]

**Comment by s-hertel** [comment]
> Given the dnf module can handle installing/uninstalling modules, and yum cannot, makes this argument moot, no? :)

Well, the `yum` action plugin may call the `dnf` module...

IMO it would be a good candidate for ansible.posix. But at least from the docs, I'd expect it to start in a community collection (maybe community.general? Probably a question for the community team), and then be promoted, since supported collections should be tried and true content.

Regarding ripping the existing support out, we could perhaps deprecate + eventually remove it. There are many cases of modules that accidentally started doing too much though, and often those are just left alone. For example, the `s3` module used to be the only module for managing AWS S3, and handled objects in buckets and buckets. Then an s3_bucket module was created to better support bucket management rather than continuing to add to the s3 module. Nothing was deprecated from the s3 module though, since in it's limited scope, it works fine and breaking playbooks was unnecessary.

ticket needs moar module :-)

**Comment by evgeni** [comment]
> > Given the dnf module can handle installing/uninstalling modules, and yum cannot, makes this argument moot, no? :)
> 
> Well, the `yum` action plugin may call the `dnf` module...

Yuck, I forgot about this…

> IMO it would be a good candidate for ansible.posix. But at least from the docs, I'd expect it to start in a community collection (maybe community.general? Probably a question for the community team), and then be promoted, since supported collections should be tried and true content.

Ack, I can work on a `dnf_module` for `community.general` (and later `ansible.posix`), but it seems like a rather long path to get it into the users hands (not your fault, just general grief)

> Regarding ripping the existing support out, we could perhaps deprecate + eventually remove it. There are many cases of modules that accidentally started doing too much though, and often those are just left alone. For example, the `s3` module used to be the only module for managing AWS S3, and handled objects in buckets and buckets. Then an s3_bucket module was created to better support bucket management rather than continuing to add to the s3 module. Nothing was deprecated from the s3 module though, since in its limited scope, it works fine and breaking playbooks was unnecessary.

My main point about ripping it out is that it's incomplete, and users usually will go "oh, there is a builtin module, that's best supported for sure, let's use it" and then it breaks on the second use.

> ticket needs moar module :-)



SCNR ;-)

**Comment by Thulium-Drake** [comment]
@evgeni o/

Any update on this? I'm actually running into this problem trying to set up Satellite 6.11 on RHEL8 with Ansible :-(

**Comment by s-hertel** [comment]
Hi!

Thanks very much for your submission to Ansible.  It means a lot to us that you've taken time to contribute. We discussed this in backlog today, and don't want to add additional top level `state` options. Rather than overloading the existing Ansible module, this would be more manageable as its own module.

However, we're absolutely always up for discussion.
Because this project is very active, we're unlikely to see comments made on closed tickets and we lock them after some time.
If you or anyone else has any further questions, please let us know by using any of the communication methods listed in the page below:

   * https://docs.ansible.com/ansible/latest/community/communication.html

In the future, sometimes starting a discussion on the development list prior to implementing a feature can make getting things included a little easier, but it's not always necessary.

Thank you once again for this and your interest in Ansible!

**Pull request closed without merging**
